### PR TITLE
Implement http api client

### LINKED ISSUE #24: implement HttpApiClient
(With regards to #23 the name of the class might be changed to ServiceClient, but for now we'll go with the original naming scheme)

# Purpose
This class will be the main entry point for all deriving packages. It's purpose is to offer optimized functions to access Api endpoints with different methods, options, file attachments etc. and returning a fully typed Promise in the end. These methods will then be used directly by a user or indirectly through a deriving package with customized parameters and return types.

Since the package is meant to be a basis for many others, it focuses on ease-of-use and developer experience more than strict patterns.

# Sample usage
Consider the following derived sample package to be the standard use-case:

```
class ProductsControllerEndpoints {
    constructor(private api: HttpApiClient) { }

    get(id: number, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<IProductData>> {
        return this.api.get(`products/${id}`, options);
    }
}

export class ShopApiClient {
    constructor(private apiClient?: ApiClient) { }

    get products() {
        return new ProductsApiProxy(this.apiClient);
    }
}
```

This may not be ideal, but working with the final class is very comfortable and allows for a great developer experience, including full intellisense on all parameters and returned values, while keeping the resulting object immutable and injectable. 

A different project would use the resulting packages like this:

```
// Setup for Singleton or DI-Container
import {ShopApiClient} from "shop-api-client";

let apiClient = (new HttpApiClient()).withBaseUrl("http://api.crazy-factory.com/");
const shopApi = new ShopApiClient(apiClient);

// Usage in a project file
import {ShopApiClient} from "shop-api-client";

let shopApi = ServiceManager.get<ShopApiClient>();
shopApi.products.get(15).then((res) => {
  // Get product name
  let name = res.data.name; 
});
```

As a bonus, the resulting packages are simple to write and therefor simple to be automatically generated. 

# Requirements
The following class interface should be updated in http-client.d.ts and implemented:

```
export class HttpApiClient extends HttpClient {
  get<T>(url: string, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>>;
  head<T>(url: string, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>>;
  delete<T>(url: string, data?: any, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>>;
  patch<T>(url: string, data?: any, files?: any[], options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>>;
  post<T>(url: string, data?: any, files?: any[], options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>>;  
  put<T>(url: string, data?: any, files?: any[], options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>>;

  request<T>(method: string, url: string, data?: any, files?: any[], options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>>;
}
```

All the above mentioned methods (except for request()) will forward their calls to request() which will validate, reformat and forward the request to the HttpClient.fetch() method and returning it's result up the stack.

Request() should fail
- on unknown method types
- if data or files are set for GET or HEAD requests
- if files are set for DELETE requests.

All requests made by this are (for now) considered to be JSON and the headers should be set accordingly (and automatically) via the default requestoptions. 

The files[] array can be seen as a functional placeholder for now. It will be implemented with 1.1.

### PR DESCRIPTION
**Pull request by dacgray**
Fixes #24

### CODE REVIEW

File: src/httpClient.ts [hunk 1]
```diff
@@ -151,3 +152,113 @@ export class HttpClientResponse<T> implements IHttpClientResponse {
...
+        if (typeof url !== "string") {
+            throw new Error("url must be a string");
+        }
+
+        if (typeof data !== "object" && typeof data !== "undefined") {
+            throw new Error("data must be an object or undefined");
+        }
+
+        if (typeof files !== "object" && typeof files !== "undefined") {
+            throw new Error("files must be an object or undefined");
```
Comment (wmathes) [inline src/httpClient.ts]: for now: Just throw an "Not implemented yet"-Error for everything that is not undefined. We will revisit this in a future issue.

Comment (wmathes) [inline src/httpClient.ts]: Verify by tests

File: src/httpClient.ts [hunk 2]
```diff
@@ -151,3 +152,113 @@ export class HttpClientResponse<T> implements IHttpClientResponse {
...
+
+        if (typeof options !== "object" && typeof options !== "undefined") {
+            throw new Error("options must be an object or undefined");
+        }
+
+        // Validate method/parameter combinations
+        switch (method) {
+            case "GET":
+                if (data || files) {
+                    throw new Error("data and files cannot be defined on a GET call");
```
Comment (wmathes) [inline src/httpClient.ts]: I know it's convenient to write it like this, but for error handling it would be good to have separate error messages for both cases. You may use fall-through cases though and can use the method variable within the error message to shorten things a bit :)

File: src/httpClient.ts [hunk 3]
```diff
@@ -151,3 +152,113 @@ export class HttpClientResponse<T> implements IHttpClientResponse {
...
+                break;
+        }
+
+        options = options || {};
+
+        options.method = method;
+        options.url = url;
+
+        if (data) {
+            options.data = data;
```
Comment (wmathes) [inline src/httpClient.ts]: logic problem. data can be all kinds of falsy values and still be correct! E.g. `false`, `null` or `0` are perfectly valid values which will not be passed on to data. It's best to just accept data as it is (because it can be undefined or virtually anything ^^)

Update your tests and implementation for this :)

File: src/httpClient.ts [hunk 4]
```diff
@@ -151,3 +152,113 @@ export class HttpClientResponse<T> implements IHttpClientResponse {
...
+
+        if (["GET", "HEAD", "DELETE", "POST", "PATCH", "PUT"].indexOf(method) < 0) {
+            throw new Error("method can be one of GET, HEAD, DELETE, PATCH, POST, PUT");
+        }
+
+        if (typeof url !== "string") {
+            throw new Error("url must be a string");
+        }
+
+        if (typeof data !== "object" && typeof data !== "undefined") {
```
Comment (wmathes) [inline src/httpClient.ts]: data can be absolutely anything. Remove this check and update your tests.

it's more important to ensure (in below's switch case) that data is definetely `undefined` for certain methods. All other methods may accept absolutely everything :)

File: test/HttpApiClientTest.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
+import {HttpApiClient, HttpClient} from "../src/httpClient";
+
+describe("HttpApiClient", () => {
+    let httpApiClient: HttpApiClient;
+
+    it("should return a valid baseUrl after construction", () => {
```
Comment (wmathes) [inline test/HttpApiClientTest.ts]: Split this test into 2 separate tests as both scenarios are quite different and the title actually may not fit the first case. How about *it should leave the baseUrl untouched if omitted during construction*

File: test/HttpApiClientTest.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+        httpApiClient = new HttpApiClient();
+        expect(httpApiClient.baseUrl).toBeUndefined();
+
+        httpApiClient = new HttpApiClient("https://myapi.crazy-factory.com");
+        expect(httpApiClient.baseUrl).toEqual("https://myapi.crazy-factory.com");
+
+    });
+
+    it("should throw errors for incorrect parameters", () => {
```
Comment (wmathes) [inline test/HttpApiClientTest.ts]: maybe rephrase that to "incorrect method" and also test for undefined/null here.

File: src/typings/http-client.d.ts [hunk 7]
```diff
@@ -52,12 +53,8 @@ declare module "http-client" {
...
     }
 
-    export class HttpApiClient {
-
-    }
-
     interface IHttpClientMiddleware {
         (config: IHttpClientRequestOptions, next: (options: IHttpClientRequestOptions) => Promise<any>):
-            void | IHttpClientRequestOptions | HttpClientResponse;
+            void | IHttpClientRequestOptions | HttpClientResponse<any>;
```
Comment (wmathes) [inline src/typings/http-client.d.ts]: The return type is wrong. This is not your error, but we should open up an issue to fix this. I believe it should just allow for Promise<HttpClientResponse<any>>. This should be changed for the return type of the *next*-parameter as well.

You can open up an issue for this and fix it within the same PullRequest.

File: src/httpClient.ts (hunk 2 above)
Comment (wmathes) [inline src/httpClient.ts]: PS: Yes you will not be able to test the files thingy with the NYI-Error further above, but this will change soon and having these cases in already is ok.

File: test/HttpApiClientTest.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+    });
+
+    it("should throw an error on incorrect GET parameters", () => {
+
+        expect(() => {
+            httpApiClient.request("GET", "/users/1", {id: 1}, undefined, undefined);
+        }).toThrowError();
+
+        expect(() => {
```
Comment (wmathes) [inline test/HttpApiClientTest.ts]: Can be a separate tests a la "should not allow files for GET requests". You can use the .get() function directly here.

Optional: For good measure (and practice) you may also wrap all tests for files acceptance into a nested describe block. Same for data. Promise returns etc.

Alternatively: You can group by method and have the separate tests within the method group. 

(Grouping is done by having a describe() within a describe())

File: test/HttpApiClientTest.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+    it("should throw an error on incorrect DELETE parameters", () => {
+
+        expect(() => {
+            httpApiClient.request("DELETE", "/users/1", undefined, [{file: "iamafile"}], undefined);
+        }).toThrowError();
+
+    });
+
+    it("should return a valid promise from get()", () => {
```
Comment (wmathes) [inline test/HttpApiClientTest.ts]: You may test this just once calling request() as all other functions will call request() anyway. You can skip the test for the individual methods.

Additionally: Make sure, that they actually call request() and return it's value in the end. You can achieve this by creating a httpApiObject and wrapping/replacing it's request() method. You can then save the return value before it's return to you by get()/post() etc. and compare the result with toBe()

Also: Expand this test to use addMiddleware() to return a Promise. There's no need to call a real URL here.

File: test/HttpApiClientTest.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            expect(res).toBeDefined();
+        });
+    });
+
+    it("should return a valid promise from post()", () => {
+        httpApiClient = new HttpApiClient();
+        httpApiClient.post("/users").then((res) => {
+            expect(res).toBeDefined();
+        });
+        httpApiClient = new HttpApiClient("https://myapi.crazy-factory.com");
```
Comment (wmathes) [inline test/HttpApiClientTest.ts]: this functionality would be handled by a Middleware of the HttpClient object itself. You can remove this part of the test :)

File: .gitignore [hunk 11]
```diff
@@ -5,5 +5,7 @@
 /node_modules/
 /report/
 /typings/
+/.idea/
+/dist/
 .DS_Store
```
Comment (wmathes) [inline .gitignore]: technically not part of the issue. 

/,idea/ has already been added in master. Rebase your branch to see the effect :) 
/dist/ will be added by fixing #21 so you may remove it as well.

File: src/httpClient.ts (hunk 1 above)
Comment (wmathes) [inline src/httpClient.ts]: Second thought... move this test below the switch case and everything should be testable just fine.

File: src/httpClient.ts [hunk 12]
```diff
@@ -151,3 +152,113 @@ export class HttpClientResponse<T> implements IHttpClientResponse {
...
     }
 }
+
+export class HttpApiClient extends HttpClient {
+
+    baseUrl: any;
+
+    constructor(baseUrl: string = undefined) {
+        super();
+        this.baseUrl = baseUrl;
```
Comment (dacgray) [inline src/httpClient.ts]: Solved by setting configuration in constructor

    constructor(baseUrl: string = undefined) {

        super();

        this.configure((config) => {
            if (baseUrl) {
                config.withBaseUrl(baseUrl);
            }
            this.configuration = config;
        });

    }

File: src/httpClient.ts (hunk 4 above)
Comment (dacgray) [inline src/httpClient.ts]: // Validate method/parameter combinations
        if (method == "GET" || method == "HEAD") {
            if (data !== undefined || files !== undefined) {
                throw new Error("data and files cannot be defined on a GET call");
            }
        }
        if (method == "DELETE") {
            if (files !== undefined) {
                throw new Error("files cannot be defined on a DELETE call");
            }
        }

File: src/httpClient.ts (hunk 2 above)
Comment (dacgray) [inline src/httpClient.ts]: TS doesn't like multiple constraints per case, so just using if statements:

     // Validate method/parameter combinations
        if (method == "GET" || method == "HEAD") {
            if (data !== undefined) {
                throw new Error("data cannot be defined on a "+method+" call");
            }
            if (files !== undefined) {
                throw new Error("files cannot be defined on a "+method+" call");
            }
        }
        if (method == "DELETE" && files !== undefined) {
            throw new Error("files cannot be defined on a DELETE call");
        }

File: src/httpClient.ts (hunk 3 above)
Comment (dacgray) [inline src/httpClient.ts]: @wmathes     should I remove the constraint, or use:

     if (data !== undefined) {
            options.data = data;
        }
        if (files !== undefined) {
            options.files = files;
        }

**Comment by dacgray** [comment]
Fixes #37

File: src/httpClient.ts [hunk 13]
```diff
@@ -176,89 +183,69 @@ export class HttpApiClient extends HttpClient {
...
-            requestConfig.options,
+            this.configuration.options,
             options);
 
         return this.fetch(url, httpClientRequestOptions);
     }
 
-    get<T>(url: string, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>> {
-        return this.request("GET", url, undefined, undefined, options);
+    get<T>(url: string, serviceClientRequestOptions?: IServiceClientRequestOptions): Promise<HttpClientResponse<T>> {
```
Comment (wmathes) [inline src/httpClient.ts]: same as above. No need for the long parameter name

File: src/httpClient.ts [hunk 14]
```diff
@@ -176,89 +183,69 @@ export class HttpApiClient extends HttpClient {
...
+        return this.request("GET", url, undefined, undefined, serviceClientRequestOptions);
     }
 
-    head<T>(url: string, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>> {
-        return this.request("HEAD", url, undefined, undefined, options);
+    head<T>(url: string, serviceClientRequestOptions?: IServiceClientRequestOptions): Promise<HttpClientResponse<T>> {
```
Comment (wmathes) [inline src/httpClient.ts]: same as above. No need for the long parameter name

File: src/httpClient.ts [hunk 15]
```diff
@@ -176,89 +183,69 @@ export class HttpApiClient extends HttpClient {
...
+        return this.request("HEAD", url, undefined, undefined, serviceClientRequestOptions);
     }
 
-    delete<T>(url: string, data?: any, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>> {
-        return this.request("DELETE", url, data, undefined, options);
+    delete<T>(url: string, data?: any, serviceClientRequestOptions?: IServiceClientRequestOptions):
```
Comment (wmathes) [inline src/httpClient.ts]: same as above. No need for the long parameter name

File: src/httpClient.ts [hunk 16]
```diff
@@ -176,89 +183,69 @@ export class HttpApiClient extends HttpClient {
...
+    Promise<HttpClientResponse<T>> {
+        return this.request("DELETE", url, data, undefined, serviceClientRequestOptions);
     };
 
-    patch<T>(url: string, data?: any, files?: any, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>> {
-        return this.request("PATCH", url, data, files, options);
+    patch<T>(url: string, data?: any, files?: any, serviceClientRequestOptions?: IServiceClientRequestOptions):
```
Comment (wmathes) [inline src/httpClient.ts]: same as above. No need for the long parameter name

File: src/httpClient.ts [hunk 17]
```diff
@@ -176,89 +183,69 @@ export class HttpApiClient extends HttpClient {
...
+    Promise<HttpClientResponse<T>> {
+        return this.request("PATCH", url, data, files, serviceClientRequestOptions);
     };
 
-    post<T>(url: string, data?: any, files?: any, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>> {
-        return this.request("POST", url, data, files, options);
+    post<T>(url: string, data?: any, files?: any, serviceClientRequestOptions?: IServiceClientRequestOptions):
```
Comment (wmathes) [inline src/httpClient.ts]: same as above. No need for the long parameter name

File: src/httpClient.ts [hunk 18]
```diff
@@ -176,89 +183,69 @@ export class HttpApiClient extends HttpClient {
...
+    Promise<HttpClientResponse<T>> {
+        return this.request("POST", url, data, files, serviceClientRequestOptions);
     };
 
-    put<T>(url: string, data?: any, files?: any, options?: IHttpClientRequestOptions): Promise<HttpClientResponse<T>> {
-        return this.request("PUT", url, data, files, options);
+    put<T>(url: string, data?: any, files?: any, serviceClientRequestOptions?: IServiceClientRequestOptions):
```
Comment (wmathes) [inline src/httpClient.ts]: same as above. No need for the long parameter name

File: src/httpClient.ts [hunk 19]
```diff
@@ -176,89 +183,69 @@ export class HttpApiClient extends HttpClient {
...
-        });
+        options.data = data;
+        options.files = files;
 
         let httpClientRequestOptions: IHttpClientRequestOptions = Object.assign(
             {},
             HttpClientConfiguration.defaults,
-            requestConfig.options,
+            this.configuration.options,
             options);
```
Comment (wmathes) [inline src/httpClient.ts]: This is redundant and mostly handled in .fetch() already.

You can just combine `options` with a `fetchOptions`-var and pass it into fetch(). Talk to me once you reach this point so i can assist :)

File: test/HttpApiClientTest.ts [hunk 20]
```diff
@@ -1,9 +1,10 @@
 import {HttpApiClient, HttpClient} from "../src/httpClient";
 
 describe("HttpApiClient", () => {
+
     let httpApiClient: HttpApiClient;
 
-    it("should return a valid baseUrl after construction", () => {
+    it("should set the baseUrl to the same value as httpClient if omitted during construction", () => {
```
Comment (wmathes) [inline test/HttpApiClientTest.ts]: This test is a duplicate?

File: src/httpClient.ts [hunk 21]
```diff
@@ -176,89 +183,69 @@ export class HttpApiClient extends HttpClient {
...
+            if (files !== undefined) {
+                throw new Error(`files cannot be defined on a ${method} call`);
+            }
+        }
+        if (method === "DELETE" && files !== undefined) {
+            throw new Error("files cannot be defined on a DELETE call");
         }
 
-        options = options || {};
+        let options: any = serviceClientRequestOptions || {};
```
Comment (dacgray) [inline src/httpClient.ts]: We need to redefine the type, or it will throw a TS2339 error because it thinks options is of typeIServiceClientRequestOptions
